vllm/v1/stats/stats_manager.py

```diff
@@ -0,0 +1,366 @@
+                stats.max_tokens_requests.append(r.sampling_params.max_tokens)
+                stats.finished_reason_requests.append(r.finish_reason)
+
+            # Update the new output token stats from the update.
+            if len(r.output_token_latency_s_lst) > len(
+                    prev_output_token_latency_s_lst):
+                stats.time_per_output_tokens_iter.extend(
+                    r.output_token_latency_s_lst[
+                        len(prev_output_token_latency_s_lst):])
+        if stats.num_generation_tokens_requests:
+            stats.max_num_generation_tokens_requests.append(
+                max(stats.num_generation_tokens_requests))
+
+    def make_stats(self, engine_core_snapshot: EngineStatsSnapshot) -> Stats:
+        """
+        Make the engine stats from the current snapshot.
+        """
+        # Computed metrics from this iteration (the current iteration is
+        # captured in the new snapshot).
+        stats = Stats(now=time.time())
+        self._build_system_stats(stats, engine_core_snapshot)
+
+        # Get the latest timestamp from the snapshot.
+        latest_snapshot_ts_s = 0
+        for req_stats_update in engine_core_snapshot.requests_stats_updates:
+            latest_snapshot_ts_s = max(latest_snapshot_ts_s,
+                                       req_stats_update.ts_s)
+
+        # Filter out the updates that are older than the latest snapshot.
+        updates_iter = []
+        updates_remained = []
+        for update in self._request_updates:
+            if update.ts_s <= latest_snapshot_ts_s:
+                updates_iter.append(update)
+            else:
+                updates_remained.append(update)
+
+        self._request_updates = updates_remained
+
+        # Merge the updates for this iteration.
+        req_updates_iter = (engine_core_snapshot.requests_stats_updates +
+                            updates_iter)
+
+        # Group by the request id.
+        req_updates_by_id: Dict[str,
+                                List[RequestStatsUpdate]] = defaultdict(list)
+        for r_update in req_updates_iter:
+            req_updates_by_id[r_update.request_id].append(r_update)
+
+        # Sort by the timestamp.
+        for _, req_updates in req_updates_by_id.items():
+            req_updates.sort(key=lambda x: x.ts_s)
+
+        self._build_requests_stats(stats, req_updates_by_id)
+
+        # Prune the finished requests.
+        finished_req_ids = set()
+        for req_id, r in self._request_stats.items():
+            if r.is_finished:
+                finished_req_ids.add(req_id)
+
+        for req_id in finished_req_ids:
+            del self._request_stats[req_id]
+
+        return stats
+
+    def log_stats(self, stats: Stats) -> None:
+        """
+        Log the engine stats.
+        """
+
+        for stat_logger in self._stat_loggers.values():
+            stat_logger.log(stats)
+
+
+class ThreadSafeEngineStatsManager(EngineStatsManager):
```

Comment:
There are not any threads in the `AsyncLLM` class right now (its just using `asyncio`. So the `Threadsafe` variant is not needed.

Reply:
Oh right - good catch. I wonder in the future we should probably isolate the stats logger in another thread since the stats logger could be pretty expensive (e.g. the prometheus logging is quite slow when it comes to histograms).